kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -233,52 +288,150 @@ class KinesisSink private (
     ()
   }
 
-  // TODO: limit max retries?
-  def sendBatch(batch: List[(ByteBuffer, String)], nextBackoff: Long = minBackoff): Unit =
-    if (batch.size > 0) {
+  /**
+   *  Max number of retries is unlimitted, so when Kinesis stream is under heavy load,
+   *  the events accumulate in collector memory for later retries. The fix for this is to use
+   *  sqs queue as a buffer and sqs2kinesis to move events back from sqs queue to kinesis stream.
+   *  Consider using sqs buffer in heavy load scenarios.
+   *
+   */
+  def sendBatch(batch: List[Event], nextBackoff: Long = minBackoff): Unit =
+    if (batch.nonEmpty) {
       log.info(s"Writing ${batch.size} Thrift records to Kinesis stream ${streamName}")
-      val putData = for {
-        p <- multiPut(streamName, batch)
-      } yield p
 
-      putData onComplete {
+      multiPut(streamName, batch).onComplete {
         case Success(s) => {
           val results = s.getRecords.asScala.toList
           val failurePairs = batch zip results filter { _._2.getErrorMessage != null }
           log.info(
             s"Successfully wrote ${batch.size - failurePairs.size} out of ${batch.size} records"
           )
-          if (failurePairs.size > 0) {
+          if (failurePairs.nonEmpty) {
             failurePairs.foreach(
               f =>
                 log.error(
                   s"Record failed with error code [${f._2.getErrorCode}] and message [${f._2.getErrorMessage}]"
                 )
             )
-            log.error(s"Retrying all failed records in $nextBackoff milliseconds...")
             val failures = failurePairs.map(_._1)
-            scheduleBatch(failures, nextBackoff)
+            val retryErrorMsg = s"Retrying all failed records in $nextBackoff milliseconds..."
+            sendToSqsOrRetryToKinesis(failures, nextBackoff)(retryErrorMsg)
           }
         }
         case Failure(f) => {
           log.error("Writing failed.", f)
-          log.error(s"Retrying in $nextBackoff milliseconds...")
-          scheduleBatch(batch, nextBackoff)
+          val retryErrorMsg = s"Retrying in $nextBackoff milliseconds..."
+          sendToSqsOrRetryToKinesis(batch, nextBackoff)(retryErrorMsg)
         }
       }
     }
 
-  private def multiPut(name: String, batch: List[(ByteBuffer, String)]): Future[PutRecordsResult] =
+  private def sendToSqsOrRetryToKinesis(
+    failures: List[Event],
+    nextBackoff: Long
+  )(
+    retryErrorMsg: String
+  ): Unit =
+    maybeSqs match {
+      case Some(sqs) =>
+        log.info(
+          s"Sending ${failures.size} events from a batch to SQS buffer queue: ${sqs.sqsBufferName}"
+        )
+        putToSqs(sqs, failures)
+        ()
```

Comment:
Don't we want to wait for the result of `putToSqs` instead of discarding the `Future`? What if all the writes to SQS fail?

Reply:
The batch call to SQS is retried up to 10 times. This is already the backup, so unless we want to implement some storage backup for the backup we are pretty much out of options. We don't want to keep it in memory when sqs buffer is enabled, as this is exactly the problem we're fixing with the whole sqs buffer.

Reply:
> The batch call to SQS is retried up to 10 times..
If it fails after 10 times, we must have a log message to indicate that we are dropping events.